components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt

```diff
@@ -0,0 +1,109 @@
+        // reason, the application cannot run and record them.
+        const val CRASH_FILE_NAME = "glean_crash_counts"
+
+        // These keys correspond to the labels found for crashCount metric in metrics.yaml as well
+        // as the persisted crashes in the crash count file (see above comment)
+        const val UNCAUGHT_EXCEPTION_KEY = "uncaught_exception"
+        const val NATIVE_CODE_CRASH_KEY = "native_code_crash"
+    }
+
+    private val logger = Logger("glean/GleanCrashReporterService")
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal val file = File(context.applicationInfo.dataDir, CRASH_FILE_NAME)
+
+    init {
+        // We only want to record things on the main process because that is the only one in which
+        // Glean is properly initialized.  Checking to see if we are on the main process here will
+        // prevent the situation that arises because the integrating app's Application will be
+        // re-created when prompting to report the crash, and Glean is not initialized there since
+        // it's not technically the main process.
+        if (context.isMainProcess() && file.exists() && file.canRead()) {
+            // Because lib-crash runs in a separate process when report() is called, we need to do
+            // actual recording here because the GleanCrashReporterService constructor is called and
+            // runs in the process where Glean is actually initialized.
+            //
+            // NOTE:  It is important to make sure to initialize Glean before instantiating and
+            // registering the CrashReporter!
+            val lines = file.readLines()
+
+            // It's possible that there was more than one crash recorded in the file so process each
+            // line.
+            lines.forEach { line ->
+                when (line) {
+                    UNCAUGHT_EXCEPTION_KEY -> CrashMetrics.crashCount[UNCAUGHT_EXCEPTION_KEY].add()
+                    NATIVE_CODE_CRASH_KEY -> CrashMetrics.crashCount[NATIVE_CODE_CRASH_KEY].add()
+                }
+            }
+
+            // Clear persisted counts by deleting the file
+            file.delete()
+        }
+    }
+
+    override fun report(crash: Crash.UncaughtExceptionCrash) {
+        // Persist the crash in a file so that it can be recorded on the next application start. We
+        // cannot directly record to Glean here because CrashHandler process is not the same process
+        // as Glean is initialized in.
+        // Create the file if it doesn't exist
+        if (!file.exists()) {
+            file.createNewFile()
+        }
+
+        // Add a line representing the crash that was received
+        if (file.canWrite()) {
+            file.appendText(UNCAUGHT_EXCEPTION_KEY + "\n")
+        }
+    }
+
+    override fun report(crash: Crash.NativeCodeCrash) {
+        // Persist the crash in a file so that it can be recorded on the next application start. We
```

Comment:
Are the `report` calls thread safe? Can it ever happen to have them at the same time? Can it ever happen for them to be called from different processes? What if we try to write, at the same time, from different processes?

Reply:
That's all implementation details of lib-crash...   I don't think you could have an app crash for 2 reasons at the same time.  @pocmo indicated that the report() functions are called synchronously and from lib-crash's own process, so very unlikely that they would be called at the same time, or be called from another process.

Reply:
Great! Thanks for digging into that. Do you mind adding that as a comment somewhere within this file, for future reference?